### PR TITLE
Enable casting to IamDataFrame multiple times

### PR DESCRIPTION
**Pull request by danielhuppmann**
# Please confirm that this PR has done the following:

- [x] Tests Added
- [x] Documentation Added
- [x] Description in RELEASE_NOTES.md Added

# Description of PR

I recently noticed that casting an IamDataFrame again (i.e., `IamDataFrame(IamDataFrame(<file>))` raises a non-intuitive error message.

This PR implements a simple approach to circumvent this problem, with all attributes of the first and second casting of the IamDataFrame instance pointing to the same objects. This follows the logic for pandas.DataFrames (as far as I understand it) where performing an operation on one instance has an impact also on the second (see the new test to show how this works in pyam).

An alternative approach may be to do a full copy during the double-casting.

## Use case

I wrote a script recently that cast an object to an `IamDataFrame` ([see here](https://github.com/openENTRANCE/nomenclature/blob/69703d073a6367332aab1d9d648dc01a9ec483c6/nomenclature/__init__.py#L82)) before performing a few checks and operations - this failed when the object was already an `IamDataFrame`. That function has been extended to first check before casting, but this may be a problem that other users run into.

### CODE REVIEW

File: tests/test_core.py [hunk 1]
```diff
@@ -46,6 +46,19 @@ def test_init_df_with_index(test_pd_df):
     pd.testing.assert_frame_equal(df.timeseries().reset_index(), test_pd_df)
 
 
+def test_init_from_iamdf(test_df_year):
```
Comment (znicholls) [inline tests/test_core.py:49]: Do you have a test for what happens if there are kwargs?

**Comment by Rlamboll** [comment]
I think this does what it initially set out to do, and am happy to approve it. However the one use-case that I would understand for this is 'refreshing' databases. I don't know if you're aware (@znicholls has described this as a bug before but I assumed it was intentional) but changing the data doesn't update the values extracted by calls like "df.scenarios()", so I often call df = DataFrame(df.data) to ensure everything is consistent after modifying data directly. To me it would be most intuitive if these had the same effect regarding consistency.

**Comment by danielhuppmann** [comment]
> However the one use-case that I would understand for this is 'refreshing' databases. I don't know if you're aware (@znicholls has described this as a bug before but I assumed it was intentional) but changing the data doesn't update the values extracted by calls like "df.scenarios()", so I often call df = DataFrame(df.data) to ensure everything is consistent after modifying data directly. To me it would be most intuitive if these had the same effect regarding consistency.

I'm not quite sure what you mean with "refreshing" databases - do you mean an `IamDataFrame`? Also, "everything is consistent" - this should be the case anyway if you use the proper functions like `rename()`, `convert_unit()`, `append()`.

`df.scenarios()` returns a list of scenarios at the time when you call the function - it is not intended to be updated when `df` is changed...

Maybe bring this into a new issue if it should be discussed more thoroughly?

About the use case, I added a section to the PR description.

**Comment by Rlamboll** [comment]
> - this should be the case anyway if you use the proper functions like rename(), convert_unit(), append().
I don't, I directly operate on the .data and then make it consistent afterwards. I don't think it's necessarily a problem, just something to be aware of. I will start a "bug report" for clarity anyway.

File: tests/test_core.py (hunk 1 above)
Comment (danielhuppmann) [reply tests/test_core.py:49]: added a test!

**Comment by danielhuppmann** [comment]
@znicholls added a test and an additional paragraph in the notes on the to-be-expected behavior. Please say if that is sufficiently clear!

File: tests/test_core.py [hunk 2]
```diff
@@ -46,6 +46,25 @@ def test_init_df_with_index(test_pd_df):
...
+    assert all(test_df_year.scenarios().values == ['scen_b', 'scen_foo'])
+
+    # overwrites on the new object do not have effects on the original object
+    df = df.rename(scenario={'scen_foo': 'scen_bar'})
+    assert all(df.scenarios().values == ['scen_b', 'scen_bar'])
+    assert all(test_df_year.scenarios().values == ['scen_b', 'scen_foo'])
+
+
+def test_init_from_iamdf_raises(test_df_year):
+    # casting an IamDataFrame instance again with extra args fails
```
Comment (znicholls) [inline tests/test_core.py:64]: should there be any error message in particular here? I can imagine that getting a ValueError just cause you added kwargs would be confusing? Maybe it should say, "If you init with an IamDataFrame, you cannot use kwargs"?

File: pyam/core.py [hunk 3]
```diff
@@ -90,9 +90,22 @@ class IamDataFrame(object):
     kwargs :code:`sheet_name` ('data') and :code:`meta_sheet_name` ('meta')
     Calling the class with :code:`meta_sheet_name=False` will
     skip the import of the 'meta' table.
+
+    When initializing an :class:`IamDataFrame` from an object that is already
+    an :class:`IamDataFrame` instance, the new object will be hard-linked to
+    all attributes of the original object - so any changes on one object
+    (e.g., with :code:`inplace=True`) may also modify the other object!
```
Comment (znicholls) [inline pyam/core.py:97]: Maybe add that 'This is intended behaviour and consistent with Pandas but may be confusing for those who are not used to the Pandas/Python universe.'

File: tests/test_core.py (hunk 2 above)
Comment (danielhuppmann) [reply tests/test_core.py:64]: added an explicit error message (leaving the door open for valid keyword arguments in the future)

File: pyam/core.py (hunk 3 above)
Comment (danielhuppmann) [reply pyam/core.py:97]: done

**Comment by danielhuppmann** [comment]
@gidden, do you mind taking a look?

**Pull request merged**